packages/message/src/DeadLetterMessage.ts

```diff
@@ -0,0 +1,54 @@
+import { SQSRecord, EventBridgeEvent } from 'aws-lambda';
+
+import { parseSQSMessageBody } from '@cumulus/aws-client/SQS';
+import { CumulusMessage } from '@cumulus/types/message';
+import Logger from '@cumulus/logger';
+
+import { getCumulusMessageFromExecutionEvent } from './StepFunctions';
+
+const log = new Logger({ sender: '@cumulus/DeadLetterMessage' });
+
+type StepFunctionEventBridgeEvent = EventBridgeEvent<'Step Functions Execution Status Change', { [key: string]: string }>;
+type UnwrapDeadLetterCumulusMessageReturnType = (
+  StepFunctionEventBridgeEvent
+  | AWS.SQS.Message
+  | SQSRecord
+  | CumulusMessage
+);
+
+/**
+ * Unwrap dead letter Cumulus message, which may be wrapped in a
+ * States cloudwatch event, which is wrapped in an SQS message.
+ *
+ * @param {Object} messageBody - received SQS message
+ * @returns {Object} the cumulus message or nearest available object
+ */
+export const unwrapDeadLetterCumulusMessage = async (
+  messageBody: UnwrapDeadLetterCumulusMessageReturnType
+): Promise<UnwrapDeadLetterCumulusMessageReturnType> => {
+  try {
+    if ('cumulus_meta' in messageBody) {
+      return messageBody;
+    }
+    if ('Body' in messageBody || 'body' in messageBody) {
+      // AWS.SQS.Message/SQS.Record case
+      const unwrappedMessageBody = parseSQSMessageBody(
+        messageBody
+      ) as CumulusMessage;
+      return await unwrapDeadLetterCumulusMessage(unwrappedMessageBody);
+    }
+    if (!('detail' in messageBody)) {
+      // Non-typed catchall
+      return messageBody;
+    }
```

Comment:
I'm not sure, but is this logic doing the same thing as the previous iteration of the code?

```
    if (messageBody.detail !== undefined) {
      // AWS States event case
      const { input, output } = messageBody.detail;
      const unwrappedMessageBody = JSON.parse(output || input);
      return unwrapDeadLetterCumulusMessage(unwrappedMessageBody);
    }
```

Reply:
Is `this` referring to the L40-42 block?    If so, not exactly.

This is maintaining this logic:

```
 // indeterminate, possibly malformed case
    return messageBody;
```

The logic listed above is covered with:

```
    // StepFunctionEventBridgeEvent case
    const unwrappedMessageBody = await getCumulusMessageFromExecutionEvent(messageBody);
    return await unwrapDeadLetterCumulusMessage(unwrappedMessageBody);
```

and some of what was written before is covered with the logic in `getCumulusMessageFromExecutionEvent`.   

This is the second change in the PR, which deals with the <message> not having the proper values set as the db write lambda had not executed this code and populated the message before it was put on the DLQ/S3 datastore.